src/main/kotlin/org/jitsi/nlj/rtp/codec/vp9/Vp9Packet.kt

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright @ 2018 - present 8x8, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jitsi.nlj.rtp.codec.vp9
+
+import org.jitsi.nlj.RtpEncodingDesc
+import org.jitsi.nlj.RtpLayerDesc
+import org.jitsi.nlj.rtp.ParsedVideoPacket
+import org.jitsi.rtp.extensions.bytearray.hashCodeOfSegment
+import org.jitsi.utils.logging2.createLogger
+import org.jitsi.utils.logging2.cwarn
+import org.jitsi_modified.impl.neomedia.codec.video.vp9.DePacketizer
+
+/**
+ * If this [Vp9Packet] instance is being created via a clone,
+ * we've already parsed the packet itself and determined whether
+ * or not its a keyframe and what its spatial layer index is,
+ * so the constructor allows passing in those values if
+ * they're already known.  If they're null, this instance
+ * will do the parsing itself.
+ */
+class Vp9Packet private constructor (
+    buffer: ByteArray,
+    offset: Int,
+    length: Int,
+    isKeyframe: Boolean?,
+    isStartOfFrame: Boolean?,
+    isEndOfFrame: Boolean?,
+    encodingIndex: Int?,
+    pictureId: Int?,
+    TL0PICIDX: Int?
+) : ParsedVideoPacket(buffer, offset, length, encodingIndex) {
+
+    constructor(
+        buffer: ByteArray,
+        offset: Int,
+        length: Int
+    ) : this(buffer, offset, length,
+        isKeyframe = null,
+        isStartOfFrame = null,
+        isEndOfFrame = null,
+        encodingIndex = null,
+        pictureId = null,
+        TL0PICIDX = null
+    )
+
+    override val isKeyframe: Boolean =
+        isKeyframe ?: DePacketizer.VP9PayloadDescriptor.isKeyFrame(this.buffer, payloadOffset, payloadLength)
+
+    override val isStartOfFrame: Boolean =
+        isStartOfFrame ?: DePacketizer.VP9PayloadDescriptor.isStartOfFrame(buffer, payloadOffset, payloadLength)
+
+    override val isEndOfFrame: Boolean =
+        isEndOfFrame ?: DePacketizer.VP9PayloadDescriptor.isEndOfFrame(buffer, payloadOffset, payloadLength)
+
+    override val layerId: Int
+        get() = if (hasLayerIndices) RtpLayerDesc.getIndex(0, spatialLayerIndex, temporalLayerIndex) else super.layerId
+
+    /** End of VP9 picture is the marker bit. Note frame/picture distinction. */
+    /* TODO: not sure this should be the override from [ParsedVideoPacket] */
```

Comment:
Do we need to resolve this TODO?

Reply:
I don't think anything in `ParsedVideoPacket` is using it currently, so maybe the answer is just remove it from there?

Reply:
I think we could organize those logically, rather than by their current use.

Reply:
The tricky thing is that `isStartOfFrame` and `isEndOfFrame` end up having subtlely different semantics for VP8 and VP9, because for VP9 "frame" means "encoded frame" and not "picture".

The only thing used from `ParsedVideoFrame` currently is `isKeyframe`.

Reply:
Hmm, I think it's possible that, even though the semantics are different, the presence of the field would still work at the higher level.  But, we can tweak it later; this is fine for now.